### PR TITLE
Feature/two way scaling

### PR DESCRIPTION
**Pull request by mnarayan**
Implementing two-way centering and scaling of an input data matrix using successive normalization. 

- [x] Add basic alternating estimation
- [ ] Create a class analogous to sklearn's StandardScaler to wrap around the method `twoway_standardize`
- [ ] Add convergence checks
- [ ] Add twoway-corrcoef alternative to quic_graph_lasso
- [ ] Add example to show importance for BIC or any other sparse model selection along regularization path

### CODE REVIEW

File: inverse_covariance/clean.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+                    self.row_var_ = .0
+                else:
+                    self.row_var_ = None
+            self.row_mean_, self.row_var_, self.n_cols_seen_ = \
+                _incremental_mean_and_var(X.T, self.row_mean_,
+                                          self.row_var_,
+                                          self.n_cols_seen_
+                                          )
+
+        if self.with_std:
```
Comment (jasonlaska) [inline inverse_covariance/clean.py]: nit: prefer
```
self.row_scale_ = None
self.col_scale_ = None
if self.with_std:
   self.row_scale_ = _handle_zeros_in_scale(np.sqrt(self.row_var_))
   self.col_scale_ = _handle_zeros_in_scale(np.sqrt(self.col_var_))
```

File: inverse_covariance/clean.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+        copy = copy if copy is not None else self.copy
+        X = check_array(X, accept_sparse=None, copy=copy, warn_on_dtype=True,
+                        estimator=self, dtype=FLOAT_DTYPES)
+
+        if sparse.issparse(X):
+            print('Input is sparse')
+            raise NotImplemented(
+                "Algorithm for sparse matrices currently not supported.")
+        else:
```
Comment (jasonlaska) [inline inverse_covariance/clean.py]: don't need else if raise or return in previous block

File: inverse_covariance/clean.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        ----------
+        X : array-like, shape [n_rows, n_cols]
+            The data used to scale along the features axis.
+        y : (ignored)
+            .. deprecated:: 0.19
+               This parameter will be removed in 0.21.
+        copy : bool, optional (default: None)
+            Copy the input X or not.
+        """
+        if not isinstance(y, string_types) or y != 'deprecated':
```
Comment (jasonlaska) [inline inverse_covariance/clean.py]: remove `y` parameter from interface and remove this warning

File: inverse_covariance/clean.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+            Transformed array.
+        """
+        check_is_fitted(self, 'row_scale_')
+
+        copy = copy if copy is not None else self.copy
+        if sparse.issparse(X):
+            print('Input is sparse')
+            raise NotImplementedError(
+                'Algorithm for sparse matrices currently not supported.')
+        else:
```
Comment (jasonlaska) [inline inverse_covariance/clean.py]: no need for else after raise or return

File: inverse_covariance/clean.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+            raise NotImplementedError(
+                'Two Way standardization not reversible with accuracy')
+            X = np.asarray(X)
```
Comment (jasonlaska) [inline inverse_covariance/clean.py]: this code will never execute

File: inverse_covariance/tests/clean_test.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+                prng.randint(1, 5, size=(1, n_cols))
+    else:
+        mu = np.zeros(shape=(n_rows, n_cols))
+    var_rows = prng.lognormal(2, 1, size=(n_rows, 1))
+    sqcov_rows = np.diag(np.sqrt(var_rows))
+    var_cols = prng.lognormal(2, 1, size=(1, n_cols))
+    sqcov_cols = np.diag(np.sqrt(var_cols))
+    return mu + sqcov_rows * X * sqcov_cols
+
+
```
Comment (jasonlaska) [inline inverse_covariance/tests/clean_test.py]: Can we add some tests for two_way_standardize that test that it does what's expected, not just the interface?  Like with a fixed input and expected output.

File: inverse_covariance/tests/clean_test.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import numpy as np
+from scipy import sparse
+
+from sklearn.utils.testing import assert_raises
+
+from inverse_covariance.clean import (
+   twoway_standardize
+)
+
```
Comment (jasonlaska) [inline inverse_covariance/tests/clean_test.py]: Can we add some tests for the `TwoWayStandardScaler` with fixed inputs and expected outputs that run over a number of options and ensure it's right.  Also the standard sklearn check_estimator test (see example here: https://github.com/skggm/skggm/blob/develop/inverse_covariance/tests/common_test.py)

File: inverse_covariance/clean.py (hunk 5 above)
Comment (jasonlaska) [reply inverse_covariance/clean.py]: make sure there is a test that tests each of these methods, that would have picked this up

File: inverse_covariance/clean.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+        copy : bool, optional (default: None)
+            Copy the input X or not.
+        Returns
+        -------
+        X_tr : array-like, shape [n_samples, n_features]
+            Transformed array.
+        """
+        check_is_fitted(self, 'row_scale_')
+
+        copy = copy if copy is not None else self.copy
```
Comment (jasonlaska) [inline inverse_covariance/clean.py]: I think can be shortened to `copy = copy or self.copy`

File: inverse_covariance/clean.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (jasonlaska) [reply inverse_covariance/clean.py]: I addressed by just changing the file names.  If we want to subscope it to `clean` or `preprocessing` later we can but since this is the only one, keeping it flat for now.

[the remark this answers is not in the document]
Comment (jasonlaska) [reply inverse_covariance/clean.py]: *addressed when I push changes

File: inverse_covariance/clean.py [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+
+    def _reset(self):
+        """Reset internal data-dependent state of the scaler, if necessary.
+        __init__ parameters are not touched.
+        """
+
+        # Checking one attribute is enough, becase they are all set together
+        # in partial_fit
+        if hasattr(self, 'col_scale_'):
+            del self.row_scale_
```
[the remark this answers is not in the document]
Comment (jasonlaska) [reply inverse_covariance/clean.py]: Removing `_reset()` as it is an internal function is unused internally and the caller should create a new instance if they do not want to rely on previous state, rather than re-using the transformer.

File: inverse_covariance/clean.py [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+        """Compute the mean and std for both row and column dimensions.
+        Parameters
+        ----------
+        X : {array-like}, shape [n_rows, n_cols]
+            The data used to compute the mean and standard deviation
+            along both row and column axes
+        y : Passthrough for ``Pipeline`` compatibility. Input is ignored.
+        """
+
+        return self.partial_fit(X, y)
```
[the remark this answers is not in the document]
Comment (jasonlaska) [reply inverse_covariance/clean.py]: addressing this by removing partial_fit

File: inverse_covariance/clean.py [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+                    self.col_var_ = .0
+                else:
+                    self.col_var_ = None
+
+            self.col_mean_, self.col_var_, self.n_rows_seen_ = \
+                _incremental_mean_and_var(X, self.col_mean_,
+                                          self.col_var_,
+                                          self.n_rows_seen_
+                                          )
+            if not hasattr(self, 'n_cols_seen_'):
```
[the remark this answers is not in the document]
Comment (jasonlaska) [reply inverse_covariance/clean.py]: There is some confusion about how the transformer is used, it seems to be used as incremental refinment, single pass, but there is no loop or criteria for additional refinement.  Therefore, this check is not needed.

File: inverse_covariance/clean.py (hunk 5 above)
Comment (jasonlaska) [reply inverse_covariance/clean.py]: Seems like this should be a warning not an exception.

**Comment by jasonlaska** [comment]
Alright, I've reworked this a bit, I am going to leave some questions for you @mnarayan inline.

File: inverse_covariance/tests/two_way_standard_scaler_test.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+        mu = np.zeros(shape=(n_rows, n_cols))
+
+    var_rows = prng.lognormal(2, 1, size=(n_rows, 1))
+    sqcov_rows = np.diag(np.sqrt(var_rows))
+    var_cols = prng.lognormal(2, 1, size=(1, n_cols))
+    sqcov_cols = np.diag(np.sqrt(var_cols))
+
+    return mu + sqcov_rows * X * sqcov_cols
+
+
```
Comment (jasonlaska) [inline inverse_covariance/tests/two_way_standard_scaler_test.py]: @mnarayan please review all of the tests in the file to see if its behaving as you would expect.  You can add more tests to the parameterize dectorator if you know of good specific small examples.

File: inverse_covariance/two_way_standard_scaler.py [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+        if sparse.issparse(X):
+            raise NotImplemented(
+                "Input is sparse: Algorithm for sparse matrices currently not supported."
+            )
+
+        self.n_rows_, self.n_cols_ = np.shape(X)
+
+        # Q: This doesnt seem to actually get used in the transform, only in
+        #    the inverse transform which it sounds like we should not support.
+
```
Comment (jasonlaska) [inline inverse_covariance/two_way_standard_scaler.py]: @mnarayan The values computed in the fit/partial_fit routines do not get used in the two_way_standard_scaler routine, only in the "inverse transform" method.  Do we actually need them or are they just cribbed from sklearn?

File: inverse_covariance/two_way_standard_scaler.py [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+        X = check_array(
+            X,
+            accept_sparse=False,
+            copy=copy,
+            warn_on_dtype=True,
+            dtype=FLOAT_DTYPES,
+            estimator=self,
+        )
+        n_rows, n_cols = np.shape(X)
+        if self.n_cols_ != n_cols:
```
Comment (jasonlaska) [inline inverse_covariance/two_way_standard_scaler.py]: This check was required to pass the sklearn `check_estimator` suite.  The main idea is that we throw a value error if the dimention of the transform data is not the same as that of the fit method.  Does this seem right?

File: inverse_covariance/two_way_standard_scaler.py [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+        if sparse.issparse(X):
+            raise NotImplementedError(
+                "Input is sparse: Algorithm for sparse matrices currently not supported."
+            )
+
+        warnings.warn("Reversing two way transformation is not accurate.")
+
+        # Q: Should ^ be a warning or should we just rais here and delete the
+        #    rest of the code?
+
```
Comment (jasonlaska) [inline inverse_covariance/two_way_standard_scaler.py]: I've changed your exception to a warning and then the remaining code is the same as you had implemented.  Is it desirable to retain this method or should it be removed?

TODO: tests for this method.

File: inverse_covariance/two_way_standard_scaler.py (hunk 15 above)
Comment (mnarayan) [reply inverse_covariance/two_way_standard_scaler.py]: Yeah this sounds good.

File: inverse_covariance/two_way_standard_scaler.py (hunk 16 above)
Comment (mnarayan) [reply inverse_covariance/two_way_standard_scaler.py]: I don't understand the use case of inverse_transform really and only added what would be a reasonable one-step approximation to maintain consistency with API https://github.com/scikit-learn/scikit-learn/blob/f0ab589f/sklearn/preprocessing/data.py#L697. 

Is it preferable to just create a function that does nothing with a warning that this functionality is not supported? 

I think it is better to add approximate solution with some tests.

Comment (jasonlaska) [reply inverse_covariance/two_way_standard_scaler.py]: Then I can leave the interface, raise so that it breaks but says why, and delete the code.

Comment (jasonlaska) [reply inverse_covariance/two_way_standard_scaler.py]: I think its preferable not to provide an approximate solution if the utility is interface compatibility with a different utility-- it means maintenance for us and footguns for anyone using it.

File: inverse_covariance/two_way_standard_scaler.py (hunk 14 above)
Comment (jasonlaska) [reply inverse_covariance/two_way_standard_scaler.py]: @mnarayan did you see this question about partial fit?  If these variables are not needed, we also don't need partial_fit, can simplify everything, and have this be a transform-only transformer.

**Comment by mnarayan** [comment]
There is a use case for fit/partial_fit that are not a part of transform.
For instance, I want to know the row/column means in one dataset like
training samples that I later want to use to transform the test set, I may
want to retrieve the mean/variance parameters only without applying
transform.

**Comment by jasonlaska** [comment]
Ok so then as is, it seems good.  @mnarayan in order to complete this branch I need the following things from you:

[ ] added examples using this transformer
[ ] additional tests on the main transform function that test for correctness
[ ] final review of whats here

You should be able to just `git pull` to your local branch.  If there are conflicts, maybe just clone a fresh one in a new directory.

Be sure to run `pip install black` (in python3), `black inverse_covariance/` and `black examples/` before pushing or the Travis tests will not pass. This will run the autoformatter.  After that I'll take a look and merge it down.
